jobs/base_model_validator_test.py

```diff
@@ -0,0 +1,300 @@
+
+    def test_model_invalid_id_error(self):
+        model = MockModel(
+            id='123@?!*',
+            created_on=self.year_ago,
+            last_updated=self.now
+        )
+        error = errors.ModelInvalidIdError(model)
+
+        msg = (
+            'Entity id %s: Entity id does not match regex pattern'
+            % (model.id))
+
+        assert error.message == msg
+
+    def test_model_expired_error(self):
+        model = MockModel(
+            id='123',
+            deleted=True,
+            created_on=self.year_ago,
+            last_updated=self.year_ago
+        )
+
+        error = errors.ModelExpiredError(model)
+
+        msg = (
+            'Entity id %s: model marked as deleted is older than %s days'
+            % (model.id, errors.PERIOD_TO_HARD_DELETE_MODEL_IN_DAYS))
+
+        assert error.message == msg
+
+
+class BaseValidatorDoFnTests(unittest.TestCase):
+    def setUp(self):
+        self.base_validator_fn = base_model_validator.BaseValidatorDoFn()
+
+    def test_clone_model(self):
+        model = base_models.BaseModel(id='123', deleted=True)
+        clone = self.base_validator_fn.clone_model(model)
+
+        self.assertEqual(model.id, clone.id)
+        self.assertEqual(model, clone)
+        self.assertIsNot(model, clone)
+        self.assertIsInstance(clone, base_models.BaseModel)
+
+    def test_clone_with_changes(self):
+        model = base_models.BaseModel(id='123', deleted=True)
+        clone = self.base_validator_fn.clone_model(model, deleted=False)
+
+        self.assertNotEqual(model, clone)
+        self.assertIsNot(model, clone)
+        self.assertIsInstance(clone, base_models.BaseModel)
+        self.assertTrue(model.deleted)
+        self.assertFalse(clone.deleted)
+
+    def test_clone_sub_class(self):
+        class DerivedModel(base_models.BaseModel):
+            """Simple model with an extra 'field' string property."""
+
+            field = datastore_services.StringProperty()
+
+        model = DerivedModel(id='123', field='original')
+        clone = self.base_validator_fn.clone_model(model)
+
+        self.assertEqual(model, clone)
+        self.assertIsNot(model, clone)
+        self.assertIsInstance(clone, DerivedModel)
+        self.assertEqual(model.field, 'original')
+        self.assertEqual(clone.field, 'original')
+
+    def test_clone_sub_class_with_changes(self):
+        class DerivedModel(base_models.BaseModel):
+            """Simple model with an extra 'field' string property."""
+
+            field = datastore_services.StringProperty()
+
+        model = DerivedModel(id='123', field='original')
+        clone = self.base_validator_fn.clone_model(model, field='updated')
+
+        self.assertNotEqual(model, clone)
+        self.assertIsNot(model, clone)
+        self.assertIsInstance(clone, DerivedModel)
+        self.assertEqual(model.field, 'original')
+        self.assertEqual(clone.field, 'updated')
+
+
+class BaseModelValidatorTests(unittest.TestCase):
```

Comment:
Please use distinct classes for each validator (ValidateDeletedTests, ValidateModelTimeFields, etc.). We can make a base class, ModelValidatorTestBase, to put common setUp (like self.year_ago and self.year_later).